File: app_manager/appmanager/src/main/java/app/editors/manager/mvp/presenters/login/EnterpriseLoginPresenter.java
package app.editors.manager.mvp.presenters.login;

import android.content.Intent;

import java.util.List;

import app.editors.manager.R;
import app.editors.manager.app.App;
import app.editors.manager.managers.exceptions.UrlSyntaxMistake;
import app.editors.manager.managers.tools.RetrofitTool;
import app.editors.manager.mvp.models.account.AccountsSqlData;
import app.editors.manager.mvp.models.request.RequestPassword;
import app.editors.manager.mvp.models.request.RequestSignIn;
import app.editors.manager.mvp.models.response.ResponseCapabilities;
import app.editors.manager.mvp.models.response.ResponsePassword;
import app.editors.manager.mvp.models.response.ResponseSignIn;
import app.editors.manager.mvp.models.user.User;
import app.editors.manager.mvp.views.login.CommonSignInView;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import lib.toolkit.base.managers.utils.StringUtils;
import moxy.InjectViewState;
import retrofit2.Response;

@InjectViewState
public class EnterpriseLoginPresenter extends BaseLoginPresenter<CommonSignInView, ResponseSignIn> {

    public static final String TAG = EnterpriseLoginPresenter.class.getSimpleName();

    public static final String TAG_DIALOG_WAITING = "TAG_DIALOG_WAITING";
    public static final String TAG_DIALOG_LOGIN_FACEBOOK = "TAG_DIALOG_LOGIN_FACEBOOK";
    public static final String TAG_DIALOG_FORGOT_PASSWORD = "TAG_DIALOG_LOGIN_FORGOT_PASSWORD";

    protected Disposable mDisposable;

    public EnterpriseLoginPresenter() {
        App.getApp().getAppComponent().inject(this);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if(mDisposable != null) {
            mDisposable.dispose();
        }
    }

    @Override
    protected void onTwoFactorAuth(boolean isPhone, AccountsSqlData sqlData) {
        super.onTwoFactorAuth(isPhone, sqlData);
        getViewState().onTwoFactorAuth(isPhone, sqlData);
    }

    @Override
    protected void onGetUser(User user) {
        super.onGetUser(user);
        getViewState().onSuccessLogin();
    }

    @Override
    protected void onTwoFactorAuthApp(boolean isSecret, AccountsSqlData sqlData) {
        super.onTwoFactorAuthApp(isSecret, sqlData);
        getViewState().onTwoFactorAuthTfa(isSecret, sqlData);
    }

    @Override
    protected void onGooglePermission(Intent intent) {
        super.onGooglePermission(intent);
        getViewState().onGooglePermission(intent);
    }

    public void signInPortal(final String login, final String password, final String portal) {
//        mPreferenceTool.setLogin(login);
        if (!StringUtils.isEmailValid(login)) {
            getViewState().onEmailNameError(mContext.getString(R.string.errors_email_syntax_error));
            return;
        }

        getViewState().onWaitingDialog(mContext.getString(R.string.dialogs_sign_in_portal_header_text), TAG_DIALOG_WAITING);
        final RequestSignIn requestSignIn = new RequestSignIn();
        requestSignIn.setUserName(login);
        requestSignIn.setPassword(password);
        signIn(requestSignIn, portal);
    }

    public void checkProviders() {
        final String portal = mPreferenceTool.getPortalFullPath();
        if (portal != null && !portal.isEmpty()) {
            try {
                new RetrofitTool(mContext)
                        .setSslOn(mPreferenceTool.getSslState())
                        .setCiphers(mPreferenceTool.getSslCiphers())
                        .init(portal)
                        .getApi("")
                        .capabilities().enqueue(new CommonCallback<ResponseCapabilities>() {
                    @Override
                    public void onSuccessResponse(Response<ResponseCapabilities> response) {
                        final List<String> providers = response.body().getResponse().getProviders();
                        if (providers != null && !providers.isEmpty()) {
                            getViewState().showGoogleLogin(providers.contains("google"));
                            getViewState().showFacebookLogin(providers.contains("facebook"));
                        }
                    }
                });
            } catch (UrlSyntaxMistake urlSyntaxMistake) {
                urlSyntaxMistake.printStackTrace();
            }
        }
    }

    public void checkEmail(String email) {
        if(!StringUtils.isEmailValid(email)) {
            getViewState().onError(mContext.getString(R.string.errors_email_syntax_error));
        } else {
            sendEmailNotification(email);
        }
    }

    private void sendEmailNotification(String email) {

        RequestPassword requestPassword = new RequestPassword();
        requestPassword.setPortal(mPreferenceTool.getPortal());
        requestPassword.setEmail(email);

        mDisposable = mRetrofitTool.getApiWithPreferences().forgotPassword(requestPassword)
                .map(ResponsePassword::getResponse)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(message -> {
                    getViewState().onSuccessSendEmail(message);
                }, throwable -> {
                    getViewState().onError(throwable.getMessage());
                });

    }

}

File: app_manager/appmanager/src/main/java/app/editors/manager/mvp/presenters/login/EnterpriseLoginPresenter.kt
package app.editors.manager.mvp.presenters.login

import android.content.Intent
import android.util.Log
import app.documents.core.model.cloud.CloudAccount
import app.documents.core.model.cloud.CloudPortal
import app.documents.core.model.login.request.RequestSignIn
import app.documents.core.network.common.NetworkResult
import app.documents.core.network.common.utils.SocialSignIn
import app.documents.core.network.common.utils.SocialUtils
import app.editors.manager.R
import app.editors.manager.app.App
import app.editors.manager.mvp.views.login.CommonSignInView
import kotlinx.coroutines.launch
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import lib.toolkit.base.managers.utils.StringUtils
import moxy.InjectViewState
import moxy.presenterScope

@InjectViewState
open class EnterpriseLoginPresenter : BaseLoginPresenter<CommonSignInView>() {

    companion object {
        val TAG: String = EnterpriseLoginPresenter::class.java.simpleName
        const val TAG_DIALOG_WAITING = "TAG_DIALOG_WAITING"
        const val TAG_DIALOG_LOGIN_FACEBOOK = "TAG_DIALOG_LOGIN_FACEBOOK"
    }

    private val twitterRepository
        get() = App.getApp().loginComponent.twitterLoginRepository

    val currentPortal: CloudPortal?
        get() = App.getApp().loginComponent.currentPortal

    var useLdap: Boolean = false

    var socialSignIn: SocialSignIn? = null
        private set

    init {
        App.getApp().appComponent.inject(this)
    }

    override fun onTwoFactorAuth(phoneNoise: String?, request: RequestSignIn) {
        super.onTwoFactorAuth(phoneNoise, request)
        viewState.onTwoFactorAuth(phoneNoise, Json.encodeToString(request))
    }

    override fun onTwoFactorAuthApp(secretKey: String?, request: RequestSignIn) {
        viewState.onTwoFactorAuthTfa(secretKey, Json.encodeToString(request))
    }

    override fun onAccountCreateSuccess(account: CloudAccount) {
        super.onAccountCreateSuccess(account)
        viewState.onSuccessLogin()
    }

    override fun onGooglePermission(intent: Intent?) {
        intent?.let { viewState.onGooglePermission(it) }
    }

    fun signInPortal(login: String, password: String, portal: CloudPortal) {
        if (!useLdap && !StringUtils.isEmailValid(login)) {
            viewState.onEmailNameError(context.getString(R.string.errors_email_syntax_error))
            return
        }

        viewState.onWaitingDialog()
        if (App.getApp().loginComponent.currentPortal == null) {
            App.getApp().refreshLoginComponent(portal)
        }

        signInWithEmail(login, password)
    }

    fun signInPortalSocial(social: String) {
        socialSignIn = SocialUtils.getSocialSignIn(social)
        socialSignIn?.let { social ->
            if (social is SocialSignIn.Twitter) {
                viewState.onWaitingDialog()
                signInJob = presenterScope.launch {
                    twitterRepository.getRequestToken(
                        consumerKey = social.clientId,
                        consumerSecret = social.SECRET_KEY,
                        callbackUrl = social.callbackUrl
                    ).collect { result ->
                        if (result is NetworkResult.Success) {
                            viewState.onSocialAuth(social.authUrl, result.data)
                        }
                        if (result is NetworkResult.Error) {
                            fetchError(result.exception)
                        }
                    }
                }
            } else {
                viewState.onSocialAuth(socialSignIn)
            }
        }
    }

    fun requestAccessToken(oauthToken: String, oauthVerifier: String){
        presenterScope.launch {
            twitterRepository.getAccessToken(oauthToken, oauthVerifier).collect { result ->
                if (result is NetworkResult.Success) {
                    //TODO
                    Log.d("TAG", "oauthToken = ${result.data.oauthToken}")
                    Log.d("TAG", "oauthTokenSecret = ${result.data.oauthTokenSecret}")
                }
            }
        }
    }
}